src/content/docs/tutorials/crates-tui/events.md

```diff
@@ -0,0 +1,59 @@
+---
+title: Events
+---
+
+We've already discussed `Events` and an `EventHandler` extensively in the
+[counter app](../counter-app/multiple-files/event). And you can use the exact same approach in your
+`async` application to the same effect.
+
+However, when using `tokio`, we have a few more options available to us. In this tutorial, we can
+take advantage of [`tokio_stream`] to create custom streams and fuse them together to get async
+events.
+
+[`tokio_stream`]: https://docs.rs/tokio-stream/latest/tokio_stream/
+
+First, we will create a `Event` enum:
+
+```rust
+{{#include @code/crates-tui-tutorial-app/src/events.rs:event}}
+```
+
+This will represent all possible events we can receive from the terminal.
+
+Next we can create a `crossterm_stream`:
+
+```rust
+type Stream = Pin<Box<dyn futures::Stream<Item = Event>>>;
+
+{{#include @code/crates-tui-tutorial-app/src/events.rs:crossterm}}
+```
+
+We can create another stream using an `IntervalStream` for generating `Event::Tick` events.
+
+```rust
+{{#include @code/crates-tui-tutorial-app/src/events.rs:tick}}
+```
+
+And we can do the same for `Event::Render` as well:
+
+```rust
+{{#include @code/crates-tui-tutorial-app/src/events.rs:render}}
+```
```

Comment:
We don't need either of these streams in this app

Reply:
I've removed `Tick` but I think we still want `Render`?

Reply:
What for?

Reply:
Without it, presumably you'd want to add a `self.draw` inside the `run()`'s loop, right?

But that won't work because I believe this blocks indefinitely if there are not other events:

```
            if let Some(e) = events.next().await {
                self.handle_event(e)?.map(|action| self.tx.send(action));
            }
```

Here's a video showing the error, where it only renders when I press a key:

https://github.com/ratatui-org/website/assets/1813121/29be62ec-0c25-481b-846b-cb718d888718

Here's the diff to reproduce it:

```diff
diff --git a/code/crates-tui-tutorial-app/src/app.rs b/code/crates-tui-tutorial-app/src/app.rs
index 46cfe04..a3b7d10 100644
--- a/code/crates-tui-tutorial-app/src/app.rs
+++ b/code/crates-tui-tutorial-app/src/app.rs
@@ -20,7 +20,6 @@ pub enum Mode {
 // ANCHOR: action
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum Action {
-    Render,
     Quit,
     SwitchMode(Mode),
     ScrollDown,
@@ -78,6 +77,7 @@ impl App {
             if self.should_quit() {
                 break;
             }
+            self.draw(&mut tui)?
         }
         Ok(())
     }
@@ -86,7 +86,6 @@ impl App {
     // ANCHOR: app_handle_event
     fn handle_event(&mut self, e: Event) -> Result<Option<Action>> {
         let maybe_action = match e {
-            Event::Render => Some(Action::Render),
             Event::Key(key) => self.handle_key_event(key),
             _ => None,
         };
@@ -121,7 +120,6 @@ impl App {
     // ANCHOR: app_handle_action
     fn handle_action(&mut self, action: Action, tui: &mut Tui) -> Result<()> {
         match action {
-            Action::Render => self.draw(tui)?,
             Action::Quit => self.quit(),
             Action::SwitchMode(mode) => self.switch_mode(mode),
             Action::ScrollUp => self.scroll_up(),
diff --git a/code/crates-tui-tutorial-app/src/events.rs b/code/crates-tui-tutorial-app/src/events.rs
index a7acabc..f806d5a 100644
--- a/code/crates-tui-tutorial-app/src/events.rs
+++ b/code/crates-tui-tutorial-app/src/events.rs
@@ -28,7 +28,7 @@ impl Events {
     pub fn new() -> Self {
         Self {
             streams: tokio_stream::StreamMap::from_iter([
-                ("render", render_stream()),
+                // ("render", render_stream()),
                 ("crossterm", crossterm_stream()),
             ]),
         }
```

Reply:
I think the difference is whether you're writing a game style event loop, or a gui style loop.
In a gui style loop, nothing renders unless something changed
In a game style loop, time changing is pretty relevent as things are always moving

If we're going for the gui style, which seems logical for a terminal app, and it's one where we're doing async things, then events that run in the background should generally be able to send a Redraw action pretty easily. `self.tx.send(Action::Redraw)` should work from pretty much any struct. It's a shame that the events and action sending are separate things...

And yep - I think it makes sense to move the redraw handling out of the handle_action method as it's kind of a special thing.

Reply:
> I'm not sure I follow. It is not possible to move the draw outside, right?

https://github.com/ratatui-org/crates-tui/blob/main/src/app.rs#L132-L134

Reply:
Fixed in https://github.com/ratatui-org/website/pull/440/commits/f7b4df2527b53933f865619ae622a69677b717a0

Reply:
I've removed `Action::Render`, but the `render` stream still exists.